feat(byo-pki): allow users to define their own trusted root

jvanz opened the pull request:
## Description

Adds a new `trust-config` flag to kwctl commands, allowing users to specify a path to a JSON file with the Sigstore
"application/vnd.dev.sigstore.clienttrustconfig.v0.1+json" mediaType. This ClientTrustConfig type stores the trust root configuration used to connect to a third-party Sigstore infrastructure.

With this feature, users can verify policies using their own Sigstore infrastructure instead of the upstream Sigstore components.

The new CLI flag is optional; if not specified, kwctl will continue to use the default upstream Sigstore configuration.

Fix kubewarden/kwctl#1245 


## Test

To test this feature I've used the [sigstore/scaffolding](https://github.com/sigstore/scaffolding/tree/d0a102b3edf322c9658c0e5abe94a671f49e8245) repository. It has helper scripts and documentation to spin up a full Sigstore infrastructure locally in a Kubernetes cluster. To allow you to test it as well. This is the steps:

1. Choose two Kubewarden policies artifacts to sign with cosign. One to use the upstream Sigstore infra and another to use the local infra
2. Sign and verify one of the artifacts with upstream sigstore. Something like this:
```bash
cosign sign ghcr.io/jvanz/policies/testing-sigstore-upstreaminfra:latest  

cosign verify ghcr.io/jvanz/policies/testing-sigstore-upstreaminfra:latest --certificate-identity-regexp 'https://github.com/jvanz/*' --certificate-oidc-issuer https://token.actions.githubusercontent.com

```
3. Follow the [getting started](https://github.com/sigstore/scaffolding/blob/d0a102b3edf322c9658c0e5abe94a671f49e8245/getting-started.md) documentation to launch the Sigstore infra locally
4. Sign and verify the second artifact using the local sigstore artifact
```bash
cosign sign --rekor-url $REKOR_URL --fulcio-url $FULCIO_URL --yes --allow-insecure-registry ghcr.io/jvanz/policies/testing-sigstore-localinfra:latest  --identity-token $(curl -s $ISSUER_URL)

cosign verify --rekor-url $REKOR_URL --allow-insecure-registry ghcr.io/jvanz/policies/testing-sigstore-localinfra:latest  --certificate-identity=https://kubernetes.io/namespaces/default/serviceaccounts/default --certificate-oidc-issuer=https://kubernetes.default.svc.cluster.local

```
>[!NOTE]
> All the variables used here are exported in the getting started documention from the Sigstore repository. 

5. Build the trusted root configuration:
```bash
cosign trusted-root create \
    --fulcio="url=$FULCIO_URL,certificate-chain=/home/jvanz/SUSE/scaffolding/fulcio.pem" \
    --rekor="url=$REKOR_URL,public-key=/home/jvanz/SUSE/scaffolding/rekor.pub,start-time=2024-01-01T00:00:00Z" \
	--tsa="url=$TSA_MIRROR,certificate-chain=/home/jvanz/SUSE/scaffolding/tsa.pem" \
	--ctfe="url=$CTLOG_MIRROR,public-key=/home/jvanz/SUSE/scaffolding/ctfe.pub,start-time=2024-01-01T00:00:00Z" \
```
>[!NOTE]
> All the variables used here are exported in the getting started documention from the Sigstore repository. Furthermore, the certificates and keys used in this command should be extracted from the Kubernetes secrets running Sigstore locally. You can extract the files with these commands:
> ```bash
> kubectl get secret -n tuf-system fulcio-pub-key -o json | jq -r ".data.cert" | base64 -d > fulcio.pem
> kubectl get secret -o json -n tuf-system rekor-pub-key | jq -r ".data.public" | base64 -d > rekor.pub
> kubectl get secret -o json -n tuf-system tsa-cert-chain | jq -r '.data."cert-chain"'  | base64 -d > tsa.pem
> kubectl get secret -o json -n tuf-system ctlog-public-key | jq -r ".data.public" | base64 -d > ctfe.pub
> ```

6. Build the ClientTrusConfig:
```json
{
  "mediaType": "application/vnd.dev.sigstore.clienttrustconfig.v0.1+json",
  "trustedRoot": { PASTE HERE THE OUTPUT OF THE PREVIOUS STEP }
}

```
7. Build the verification config file to verify signature data. See the step script for examples, use kwctl scaffold command or read the [Kubewarden docs](https://docs.kubewarden.io/howtos/security-hardening/secure-supply-chain#keyless-signature-validation)
8. Verify policies with kwctl. I've wrote a script to do that in my testing environment

```bash
set -x 

cat ./verification-config-local
cat ./verification-config-upstream
cat ./trust_config.json

echo "###### Verifying policies without trust config ######"
./target/release/kwctl --no-color verify --verification-config-path ./verification-config-upstream registry://ghcr.io/jvanz/policies/testing-sigstore-upstreaminfra:latest
./target/release/kwctl --no-color verify --verification-config-path ./verification-config-upstream registry://ghcr.io/jvanz/policies/testing-sigstore-localinfra:latest
./target/release/kwctl --no-color verify --verification-config-path ./verification-config-local registry://ghcr.io/jvanz/policies/testing-sigstore-upstreaminfra:latest
./target/release/kwctl --no-color verify --verification-config-path ./verification-config-local registry://ghcr.io/jvanz/policies/testing-sigstore-localinfra:latest

echo "###### Verifying policies with trust config ######"
./target/release/kwctl --no-color verify --trust-config ./trust_config.json --verification-config-path ./verification-config-upstream registry://ghcr.io/jvanz/policies/testing-sigstore-upstreaminfra:latest
./target/release/kwctl --no-color verify --trust-config ./trust_config.json --verification-config-path ./verification-config-upstream registry://ghcr.io/jvanz/policies/testing-sigstore-localinfra:latest
./target/release/kwctl --no-color verify --trust-config ./trust_config.json --verification-config-path ./verification-config-local registry://ghcr.io/jvanz/policies/testing-sigstore-upstreaminfra:latest
./target/release/kwctl --no-color verify --trust-config ./trust_config.json --verification-config-path ./verification-config-local registry://ghcr.io/jvanz/policies/testing-sigstore-localinfra:latest

```

This are the output

```
+ cat ./verification-config-local
# Default Kubewarden verification config
#
# With this config, the only valid policies are those signed by Kubewarden
# infrastructure.
#
# This config can be saved to its default location (for this OS) with:
#   kwctl scaffold verification-config > /home/jvanz/.config/kubewarden/verification-config.yml
#
# Providing a config in the default location enables Sigstore verification.
# See https://docs.kubewarden.io/next/howtos/security-hardening/secure-supply-chain
# for more Sigstore verification options.
apiVersion: v1
allOf:
- kind: genericIssuer
  issuer: https://kubernetes.default.svc.cluster.local
  subject:
    equal: https://kubernetes.io/namespaces/default/serviceaccounts/default
anyOf: null


+ cat ./verification-config-upstream
# Default Kubewarden verification config
#
# With this config, the only valid policies are those signed by Kubewarden
# infrastructure.
#
# This config can be saved to its default location (for this OS) with:
#   kwctl scaffold verification-config > /home/jvanz/.config/kubewarden/verification-config.yml
#
# Providing a config in the default location enables Sigstore verification.
# See https://docs.kubewarden.io/next/howtos/security-hardening/secure-supply-chain
# for more Sigstore verification options.
apiVersion: v1
allOf:
- kind: genericIssuer
  issuer: https://github.com/login/oauth
  subject:
    equal: <<REDACTED EMAIL>>
anyOf: null


+ cat ./trust_config.json
{
  "mediaType": "application/vnd.dev.sigstore.clienttrustconfig.v0.1+json",
  "trustedRoot": {
    "mediaType": "application/vnd.dev.sigstore.trustedroot+json;version=0.1",
    "tlogs": [
      {
        "baseUrl": "http://rekor.rekor-system.svc:8080",
        "hashAlgorithm": "SHA2_256",
        "publicKey": {
          "rawBytes": "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEAG7HfnnxnW949IuyL/IefWDm6LPeQuA9GSCJfme74tAh68H/9YonNwW8BJ0QZr00ntt5wCg9aaj7Iv25nHFROA==",
          "keyDetails": "PKIX_ECDSA_P256_SHA_256",
          "validFor": {
            "start": "2024-01-01T00:00:00Z"
          }
        },
        "logId": {
          "keyId": "entYhHILwMbf6CqKtXPtiXqDUkjSVrRJk+KiR1mKiII="
        }
      }
    ],
    "certificateAuthorities": [
      {
        "subject": {
          "organization": "Linux Foundation"
        },
        "uri": "http://fulcio.fulcio-system.svc:8080",
        "certChain": {
          "certificates": [
            {
              "rawBytes": "MIICNjCCAd2gAwIBAgIIenk7+Dh/rCgwCgYIKoZIzj0EAwIwfjEMMAoGA1UEBhMDVVNBMRMwEQYDVQQIEwpDYWxpZm9ybmlhMRYwFAYDVQQHEw1TYW4gRnJhbmNpc2NvMRYwFAYDVQQJEw01NDggTWFya2V0IFN0MQ4wDAYDVQQREwU1NzI3NDEZMBcGA1UEChMQTGludXggRm91bmRhdGlvbjAeFw0yNTExMTAxNDAzNDVaFw0yNjExMTAxNDAzNDVaMH4xDDAKBgNVBAYTA1VTQTETMBEGA1UECBMKQ2FsaWZvcm5pYTEWMBQGA1UEBxMNU2FuIEZyYW5jaXNjbzEWMBQGA1UECRMNNTQ4IE1hcmtldCBTdDEOMAwGA1UEERMFNTcyNzQxGTAXBgNVBAoTEExpbnV4IEZvdW5kYXRpb24wWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAAQeiBEHp5hGO/y82NqhTqyyCAMRvGz/ArMiBawi75ngddHKwboKoQdNm/e3FHVyGQZkt2ipMrn7GFfuA9JfzDWzo0UwQzAOBgNVHQ8BAf8EBAMCAQYwEgYDVR0TAQH/BAgwBgEB/wIBATAdBgNVHQ4EFgQURKzY1DVWDz5V//3Of1RgX2ZPZNowCgYIKoZIzj0EAwIDRwAwRAIgUapirimk73s9ohqHi1ds0MvoUiXuOa0/rk7M7Ld0htICIGa4nxUXOZrIKJoNgJSu2Np/soEW8D/R2LV6zixauFWA"
            }
          ]
        },
        "validFor": {
          "start": "2025-11-10T14:03:45Z"
        }
      }
    ],
    "ctlogs": [
      {
        "baseUrl": "http://ctlog.ctlog-system.svc:8080",
        "hashAlgorithm": "SHA2_256",
        "publicKey": {
          "rawBytes": "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEY4JI4IO+mJMnuU9WcHZFJs7lWr+SPVQixrNdEHsklIw3uXUrA2glYYxTM/0Y3UouHl5Z1plL445OlZ82R8aBDg==",
          "keyDetails": "PKIX_ECDSA_P256_SHA_256",
          "validFor": {
            "start": "2024-01-01T00:00:00Z"
          }
        },
        "logId": {
          "keyId": "yI+0HJ14Xs/ty5N0rJ854O8bQrNWptT9BKS+8RnzybQ="
        }
      }
    ],
    "timestampAuthorities": [
      {
        "subject": {
          "organization": "local",
          "commonName": "Test TSA Root"
        },
        "uri": "http://tsa.tsa-system.svc:8080",
        "certChain": {
          "certificates": [
            {
              "rawBytes": "MIIBzTCCAXKgAwIBAgIUTtIX/LH3OoxQNHeysBoxD8jMrR8wCgYIKoZIzj0EAwIwMDEOMAwGA1UEChMFbG9jYWwxHjAcBgNVBAMTFVRlc3QgVFNBIEludGVybWVkaWF0ZTAeFw0yNTExMTAxNDAxMzdaFw0zNDExMTAxNDA0MzdaMDAxDjAMBgNVBAoTBWxvY2FsMR4wHAYDVQQDExVUZXN0IFRTQSBUaW1lc3RhbXBpbmcwWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAATIbmKGiNgPSLe9vDCfKB7gc2LRswBi/jQ5DALsBVFvF4iWeg23Ul2fEr7wpwcbKYA5XU0gf2g1impSrMUBbwYCo2owaDAOBgNVHQ8BAf8EBAMCB4AwHQYDVR0OBBYEFPq64pLBh+iUx7exkRnynnhj+aGqMB8GA1UdIwQYMBaAFCeQ8xLh1xtem6eXdUrkpnWd8UVEMBYGA1UdJQEB/wQMMAoGCCsGAQUFBwMIMAoGCCqGSM49BAMCA0kAMEYCIQCV4q63q2eJUI7x2oAHZIUVQurQCMhBPAhYceU9YGk8tQIhAKtQKteSjwDIaIoL07MnZSavVJ993y9gScAVit0x3L5C"
            },
            {
              "rawBytes": "MIIB0jCCAXigAwIBAgIUI2QvaC9DPLje8kOwkb4rWECkv0kwCgYIKoZIzj0EAwIwKDEOMAwGA1UEChMFbG9jYWwxFjAUBgNVBAMTDVRlc3QgVFNBIFJvb3QwHhcNMjUxMTEwMTM1OTM3WhcNMzUxMTEwMTQwNDM3WjAwMQ4wDAYDVQQKEwVsb2NhbDEeMBwGA1UEAxMVVGVzdCBUU0EgSW50ZXJtZWRpYXRlMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEp+P7hnNiCXEfnmesx1fDVO3erQzcCBkcOgYuDURpI7G00ekSki9j8Y//LygH9c7oRbRpyw3GQHZlNGwJEyCWFqN4MHYwDgYDVR0PAQH/BAQDAgEGMBMGA1UdJQQMMAoGCCsGAQUFBwMIMA8GA1UdEwEB/wQFMAMBAf8wHQYDVR0OBBYEFCeQ8xLh1xtem6eXdUrkpnWd8UVEMB8GA1UdIwQYMBaAFJYOHWV79+SGBJnku8pEFBKJhlJGMAoGCCqGSM49BAMCA0gAMEUCIQDRwHHWthFl12BnntXW0nzHJ8j1hM+Bh0a0D/cA+hwf0AIgBw+6dtNKdMNTUdAfOYdwc/gvYFPdeg2NfZbuyK/yXeA="
            },
            {
              "rawBytes": "MIIBlDCCATqgAwIBAgIUdD6OeHGFp8prgLgQexF6p2cOqvswCgYIKoZIzj0EAwIwKDEOMAwGA1UEChMFbG9jYWwxFjAUBgNVBAMTDVRlc3QgVFNBIFJvb3QwHhcNMjUxMTEwMTM1OTM3WhcNMzUxMTEwMTQwNDM3WjAoMQ4wDAYDVQQKEwVsb2NhbDEWMBQGA1UEAxMNVGVzdCBUU0EgUm9vdDBZMBMGByqGSM49AgEGCCqGSM49AwEHA0IABPW8JureaHsT+JFo3Um8awLhVHZySn5WrBXGt6RNsURobtNwQNpqH3gC6j/1BJPM4zbaE2fT3hqxppnA4l4p/q2jQjBAMA4GA1UdDwEB/wQEAwIBBjAPBgNVHRMBAf8EBTADAQH/MB0GA1UdDgQWBBSWDh1le/fkhgSZ5LvKRBQSiYZSRjAKBggqhkjOPQQDAgNIADBFAiASw02tYxj2iYYe8ip7teMll7Eoe9KSdY06dZFsEyhwEAIhAKqiRkJbCmu0aAo8n8nhAc/7j+pF0YNEfDgpwp7pTBZR"
            }
          ]
        },
        "validFor": {
          "start": "2025-11-10T14:01:37Z"
        }
      }
    ]
  }
}
+ echo '###### Verifying policies without trust config ######'
###### Verifying policies without trust config ######
+ ./target/release/kwctl --no-color verify --verification-config-path ./verification-config-upstream registry://ghcr.io/jvanz/policies/testing-sigstore-upstreaminfra:latest
2025-11-10T19:22:55.946655Z  INFO sigstore::cosign::client_builder: Cannot parse Rekor public key with id cf1199155bddd051268d1f16ac5c0c75c009f6fb5a63f4177f8e18d7051e3fa0: Pkcs8 spki error : Ecdsa-P256 from der bytes to public key failed: unknown/unsupported algorithm OID: 1.2.840.10045.2.1
2025-11-10T19:22:59.438902Z  INFO kwctl::verify: Policy successfully verified
+ ./target/release/kwctl --no-color verify --verification-config-path ./verification-config-upstream registry://ghcr.io/jvanz/policies/testing-sigstore-localinfra:latest
2025-11-10T19:23:03.298892Z  INFO sigstore::cosign::client_builder: Cannot parse Rekor public key with id cf1199155bddd051268d1f16ac5c0c75c009f6fb5a63f4177f8e18d7051e3fa0: Pkcs8 spki error : Ecdsa-P256 from der bytes to public key failed: unknown/unsupported algorithm OID: 1.2.840.10045.2.1
2025-11-10T19:23:06.788748Z  INFO sigstore::cosign::signature_layers: Skipping OCI layer because of error error=RekorPublicKeyNotFoundError("7a7b5884720bc0c6dfe82a8ab573ed897a835248d256b44993e2a247598a8882")
Error: Policy registry://ghcr.io/jvanz/policies/testing-sigstore-localinfra:latest cannot be validated
failed to get image trusted layers: No Signature Layer passed verification

Caused by:
    No Signature Layer passed verification
+ ./target/release/kwctl --no-color verify --verification-config-path ./verification-config-local registry://ghcr.io/jvanz/policies/testing-sigstore-upstreaminfra:latest
2025-11-10T19:23:10.864568Z  INFO sigstore::cosign::client_builder: Cannot parse Rekor public key with id cf1199155bddd051268d1f16ac5c0c75c009f6fb5a63f4177f8e18d7051e3fa0: Pkcs8 spki error : Ecdsa-P256 from der bytes to public key failed: unknown/unsupported algorithm OID: 1.2.840.10045.2.1
Error: Policy registry://ghcr.io/jvanz/policies/testing-sigstore-upstreaminfra:latest cannot be validated
Image verification failed: missing signatures
The following constraints were not satisfied:
kind: genericIssuer
issuer: https://kubernetes.default.svc.cluster.local
subject: !equal https://kubernetes.io/namespaces/default/serviceaccounts/default
annotations: null

+ ./target/release/kwctl --no-color verify --verification-config-path ./verification-config-local registry://ghcr.io/jvanz/policies/testing-sigstore-localinfra:latest
2025-11-10T19:23:18.529628Z  INFO sigstore::cosign::client_builder: Cannot parse Rekor public key with id cf1199155bddd051268d1f16ac5c0c75c009f6fb5a63f4177f8e18d7051e3fa0: Pkcs8 spki error : Ecdsa-P256 from der bytes to public key failed: unknown/unsupported algorithm OID: 1.2.840.10045.2.1
2025-11-10T19:23:21.917290Z  INFO sigstore::cosign::signature_layers: Skipping OCI layer because of error error=RekorPublicKeyNotFoundError("7a7b5884720bc0c6dfe82a8ab573ed897a835248d256b44993e2a247598a8882")
Error: Policy registry://ghcr.io/jvanz/policies/testing-sigstore-localinfra:latest cannot be validated
failed to get image trusted layers: No Signature Layer passed verification

Caused by:
    No Signature Layer passed verification
+ echo '###### Verifying policies with trust config ######'
###### Verifying policies with trust config ######
+ ./target/release/kwctl --no-color verify --trust-config ./trust_config.json --verification-config-path ./verification-config-upstream registry://ghcr.io/jvanz/policies/testing-sigstore-upstreaminfra:latest
2025-11-10T19:23:25.385336Z  INFO sigstore::cosign::signature_layers: Skipping OCI layer because of error error=RekorPublicKeyNotFoundError("c0d23d6ad406973f9559f3ba2d1ca01f84147d8ffc5b8445c224f98b9591801d")
Error: Policy registry://ghcr.io/jvanz/policies/testing-sigstore-upstreaminfra:latest cannot be validated
failed to get image trusted layers: No Signature Layer passed verification

Caused by:
    No Signature Layer passed verification
+ ./target/release/kwctl --no-color verify --trust-config ./trust_config.json --verification-config-path ./verification-config-upstream registry://ghcr.io/jvanz/policies/testing-sigstore-localinfra:latest
Error: Policy registry://ghcr.io/jvanz/policies/testing-sigstore-localinfra:latest cannot be validated
Image verification failed: missing signatures
The following constraints were not satisfied:
kind: genericIssuer
issuer: https://github.com/login/oauth
subject: !equal jvanz@jvanz.com
annotations: null

+ ./target/release/kwctl --no-color verify --trust-config ./trust_config.json --verification-config-path ./verification-config-local registry://ghcr.io/jvanz/policies/testing-sigstore-upstreaminfra:latest
2025-11-10T19:23:32.272828Z  INFO sigstore::cosign::signature_layers: Skipping OCI layer because of error error=RekorPublicKeyNotFoundError("c0d23d6ad406973f9559f3ba2d1ca01f84147d8ffc5b8445c224f98b9591801d")
Error: Policy registry://ghcr.io/jvanz/policies/testing-sigstore-upstreaminfra:latest cannot be validated
failed to get image trusted layers: No Signature Layer passed verification

Caused by:
    No Signature Layer passed verification
+ ./target/release/kwctl --no-color verify --trust-config ./trust_config.json --verification-config-path ./verification-config-local registry://ghcr.io/jvanz/policies/testing-sigstore-localinfra:latest
2025-11-10T19:23:35.965221Z  INFO kwctl::verify: Policy successfully verified

```


## Additional Information

@kravciak @juadk , the scaffolding repository has a github [action](https://github.com/sigstore/scaffolding/tree/d0a102b3edf322c9658c0e5abe94a671f49e8245/actions) able to spin up a sigstore containers to allow testing on CI as well.

jvanz committed:
feat(byo-pki): allow users to define their own trusted root

Adds a new `trust-config` flag to kwctl commands, allowing users to
specify a path to a JSON file with the Sigstore
"application/vnd.dev.sigstore.clienttrustconfig.v0.1+json" mediaType.
This ClientTrustConfig type stores the trust root configuration used to
connect to a third-party Sigstore infrastructure.

With this feature, users can verify policies using their own Sigstore
infrastructure instead of the upstream Sigstore components.

The new CLI flag is optional; if not specified, kwctl will continue to
use the default upstream Sigstore configuration.

Signed-off-by: José Guilherme Vanz <jguilhermevanz@suse.com>

flavio commented on cli-docs.md:
```diff
@@ -142,6 +142,7 @@ A YAML file may contain multiple Custom Resource declarations. In this case, `kw
 * `--settings-json <VALUE>` — JSON string containing the settings for this policy
 * `-s`, `--settings-path <PATH>` — File containing the settings for this policy
 * `--sources-path <PATH>` — YAML file holding source information (https, registry insecure hosts, custom CA's...)
+* `--trust-config <PATH>` — JSON-formatted file conforming to the ClientTrustConfig message in the Sigstore protobuf specs. This file configures the entire Sigstore instance state, including the URIs used to access the CA and artifact transparency services as well as the cryptographic root of trust itself
```
I would name it `--sigstore-trust-config` to make explicit this is about sigstore

flavio commented on src/config/verification.rs:
```diff
@@ -159,7 +164,56 @@ fn build_verification_options_from_flags(
+/// provided, it uses that file to load the trust root. Otherwise, it builds the trust root from
+/// Sigstore's TUF repository.
+pub(crate) async fn build_sigstore_trust_root(
+    matches: &ArgMatches,
```
I would change the signature of the function to receive an `Option<PathBuf>`. In this way it's clear what has to be provided when invoking the function, plus testing would become easier.

flavio commented on src/config/verification.rs:
```diff
@@ -159,7 +164,56 @@ fn build_verification_options_from_flags(
+                "specified Sigstore PKI file does not exist: {}",
+                pki_file.display()
+            ));
+        }
```
I think we can skip this check, the error message provided by the next line should be explicit enough

flavio commented on src/config/verification.rs:
```diff
@@ -159,7 +164,56 @@ fn build_verification_options_from_flags(
+        }
+        let json_bytes = fs::read(pki_file).map_err(|e| {
+            anyhow!(
+                "could not read Sigstore PKI file {}: {:?}",
```
Just to stay consistent with the name used by the cli:

```suggestion
                "could not read Sigstore trust config file {}: {:?}",
```

flavio reviewed (changes requested):
Overall LGTM.

Can we have some e2e tests for that?

flavio reviewed (commented):
LGTM, I left one minor suggestion

jvanz committed:
feat(e2e): test sigstore-trust-config CLI flag.

Adds a e2e test to validate the --sigstore-trust-config CLI flag. This
test is behind a feature flag to avoid running it in a environment with
no local Sigstore running.

Signed-off-by: José Guilherme Vanz <jguilhermevanz@suse.com>

jvanz commented:
@kubewarden/kubewarden-developers I've added a e2e test to validate the `--sigstore-trust-config` CLI flag.

viccuad commented on src/config/verification.rs:
```diff
@@ -170,3 +183,42 @@ pub(crate) async fn build_sigstore_trust_root() -> Result<Option<Arc<SigstoreTru
+    })?;
+    // Sigstore.rs does not have a direct way to read from file, so we parse the
+    // protobuf spec first, then re-serialize the trusted_root field to JSON bytes
+    let client_trust_config: sigstore_protobuf_specs::dev::sigstore::trustroot::v1::ClientTrustConfig =
```
I would prefer to have this import on the top of the file, but not a blocker for me.

viccuad reviewed (approved):
Looks great! Learned about the trust root, had some questions and minor quibbles.

flavio reviewed (approved):
LGTM. Thanks a lot for having added the tests. I would create an issue about adding a docker compose file that can be used locally to run the new e2e test. Something similar to what we're doing with the policy-server and otel

jvanz commented:
> LGTM. Thanks a lot for having added the tests. I would create an issue about adding a docker compose file that can be used locally to run the new e2e test. Something similar to what we're doing with the policy-server and otel

https://github.com/kubewarden/kwctl/issues/1429
